KAFKA-5431: cleanSegments should not set length for cleanable segment files

huxihx opened the pull request:
For a compacted topic with preallocate enabled, during log cleaning, LogCleaner.cleanSegments does not have to pre-allocate the underlying file size since we only want to store the cleaned data in the file.

It's believed that this fix should also solve KAFKA-5582.

huxi-2b committed:
KAFKA-5431: LogCleaner stopped due to org.apache.kafka.common.errors.CorruptRecordException

For a compacted topic with preallocate enabled, during log cleaning, LogCleaner.cleanSegments does not have to pre-allocate the underlying file size since we only want to store the cleaned data in the file.

It's believed that this fix should also solve KAFKA-5582.

huxihx commented:
@junrao Please review this PR. Thanks.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6056/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6041/
Test PASSed (JDK 8 and Scala 2.12).

huxihx commented:
@ijuma Please take some time to review the PR. Thanks

jasonaliyetti commented on core/src/main/scala/kafka/log/LogCleaner.scala:
```diff
@@ -417,7 +417,7 @@ private[log] class Cleaner(val id: Int,
 
     val startOffset = segments.head.baseOffset
-    val records = FileRecords.open(logFile, false, log.initFileSize(), log.config.preallocate)
+    val records = FileRecords.open(logFile, false, log.initFileSize(), false)
```
Won't this lead to a fragmented file?  I'd think you'd want to preallocate this file (if enabled), and then remove all of the invalid bytes to shrink it down as is done in segment recovery.

huxihx replied:
Not exactly. The only difference here is always to create a non-preallocate target cleaned segment files. Most of the times, the cleaned files will not contain too much records, that's why we don't use pre-allocate.

junrao replied:
@huxihx : Wow. Great finding. Thanks. It does seem that we missed trimming the log in the log cleaner when preallocation is enabled. We probably want to keep FileRecords in the preallocate mode if it's enabled since it's possible for most records to be preserved after cleaning. What we can do is to follow what we do in Log.roll() by calling records.trim() before cleaned.flush(). It would also be useful to add a simple unit test  for this case (e.g., trigger a cleaning with no filtering and verify the cleaned segment size is trimmed).

ijuma commented:
It would be best for @junrao to review this, but not clear to me that avoiding pre-allocation is the right thing.

huxi-2b committed:
KAFKA-5431: cleanSegments should not set length for cleanable segment files

Trim the cleaned log segments' files before flushing them in order to save disk spaces.

huxihx commented:
@junrao @ijuma Thanks for responding. Please take some time to review it again.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6160/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6144/
Test FAILed (JDK 8 and Scala 2.12).

huxihx commented:
retest it please

junrao commented on core/src/test/scala/unit/kafka/log/LogCleanerTest.scala:
```diff
@@ -89,6 +89,35 @@ class LogCleanerTest extends JUnitSuite {
+    // append messages to the log until we have three segments
+    val random = new java.util.Random()
+    while(log.numberOfSegments < 3)
+      log.appendAsLeader(record(random.nextInt(5), log.logEndOffset.toInt), leaderEpoch = 0)
```
It seems that we jus need to create 1 log segment in the test. We just need to make sure that the size of all appended records is less than the log segment size.

junrao commented on core/src/test/scala/unit/kafka/log/LogCleanerTest.scala:
```diff
@@ -89,6 +89,35 @@ class LogCleanerTest extends JUnitSuite {
+    val stats = new CleanerStats()
+    cleaner.cleanSegments(log, segments, map, 0L, stats)
+    // the file field of FileRecords for the segment was already renamed with suffix '.deleted'
+    assertNotEquals(new File(segments.head.log.file().getAbsolutePath.replace(Log.DeletedFileSuffix, "")).length(), originalMaxFileSize)
```
Hmm, after the cleaning, the new log segment is added to the log to replace the old log segment. It seems that we don't need to replace the file name since the first log segment will be the new segment. Also, it's probably useful to add a text message in the assertNotEquals() call.

junrao reviewed (commented):
@huxihx : Thanks for the updated patch. A couple of more comments.

huxi-2b committed:
Refined LogCleanerTest.testSizeTrimmedForPreallocatedAndCompactedTopic to address Jun's comments.

huxihx commented:
@junrao Already refined the code to address your comments. Please review it again. Thank you.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6183/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6167/
Test PASSed (JDK 8 and Scala 2.12).

junrao commented on core/src/test/scala/unit/kafka/log/LogCleanerTest.scala:
```diff
@@ -89,6 +89,31 @@ class LogCleanerTest extends JUnitSuite {
+    cleaner.clean(LogToClean(new TopicPartition("test", 0), log, 2, log.activeSegment.baseOffset))
+
+    assert(log.logSegments.iterator.next.log.channel().size() < originalMaxFileSize,
+      "Cleaned segment file should be trimmed to its real size.")
```
Could we use static public void assertTrue(String message, boolean condition)?

junrao commented on core/src/test/scala/unit/kafka/log/LogCleanerTest.scala:
```diff
@@ -89,6 +89,31 @@ class LogCleanerTest extends JUnitSuite {
+    // roll the segment, so we can clean the messages already appended
+    log.roll()
+
+    // clean the log with only one message removed
```
Hmm, is the comment accurate? It seems that the cleaning will remove message at offset 0, 1 and 2?

huxihx replied:
Seems the comment is accurate. In the test the first three messages(firstDirtyOffset = 2) are keyed with 0, 1 and 0 respectively, so only the first message is regarded as 'cleanable'. Does it make sense?

junrao reviewed (commented):
@huxihx : Thanks for the new patch. A couple of more minor comments.

huxi-2b committed:
Use assertTrue instead of assert in LogCleanerTest.testSizeTrimmedForPreallocatedAndCompactedTopic

huxihx commented:
@junrao Thanks for the comments. Already changed to use `assertTrue` instead of `assert`. Please review it again.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6215/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6231/
Test PASSed (JDK 7 and Scala 2.11).

junrao commented:
@huxihx : Thanks for the latest patch. LGTM.
